cohttp/src/request.ml

```diff
@@ -110,49 +104,42 @@ let is_valid_uri path meth =
   | Some _ -> true
   | None -> not (String.length path > 0 && path.[0] <> '/')
 
-let uri { scheme; resource; headers; meth; _ } =
-  let uri =
-    match resource with
-    | "*" -> (
-        match Header.get headers "host" with
-        | None -> Uri.of_string ""
-        | Some host ->
-            let host_uri = Uri.of_string ("//" ^ host) in
-            Uri.(make ?host:(host host_uri) ?port:(port host_uri) ()))
-    | authority when meth = `CONNECT -> Uri.of_string ("//" ^ authority)
-    | path -> (
-        let uri = Uri.of_string path in
-        match Uri.scheme uri with
-        | Some _ -> (
-            Uri.(
-              (* we have an absoluteURI *)
-              match path uri with "" -> with_path uri "/" | _ -> uri))
-        | None ->
-            let empty = Uri.of_string "" in
-            let empty_base = Uri.of_string "///" in
-            let pqs =
-              match Stringext.split ~max:2 path ~on:'?' with
-              | [] -> empty_base
-              | [ path ] ->
+let uri { resource; headers; meth; _ } =
+  match resource with
+  | "*" -> (
+      match Header.get headers "host" with
+      | None -> Uri.of_string ""
+      | Some host ->
+          let host_uri = Uri.of_string ("//" ^ host) in
+          Uri.(make ?host:(host host_uri) ?port:(port host_uri) ()))
+  | authority when meth = `CONNECT -> Uri.of_string ("//" ^ authority)
+  | path -> (
+      let uri = Uri.of_string path in
+      match Uri.scheme uri with
+      | Some _ -> (
+          Uri.(
+            (* we have an absoluteURI *)
+            match path uri with "" -> with_path uri "/" | _ -> uri))
+      | None -> (
+          let empty = Uri.of_string "" in
+          let empty_base = Uri.of_string "///" in
```

Comment:
A Uri.t is a fairly heavy record - I just noticed both of these can be lifted into constants at the toplevel and not reallocated every request

Reply:
Indeed. I wonder if we should just get rid of this function altogether and just introduce separate functions to extract the Uri components that can be safely extracted out of the request.

Reply:
Probably just better to 'deallocationise' this library separately. I actually wanted to use it to test out the JS with-extensions tree to see how well local annotations would do in some places here...

Reply:
Sure, though I think this function cannot be be used in a low allocation context anyway. Many of the components this function is returning can be extracted without any allocation at all. While the `Uri.t` this function returns will always be allocated - even if it is only done once.

Reply:
Yes, it would need changing the Uri library as well... the accessors there could be more lazy. That library's design hasn't changed for about a decade -- about high time for a refresh!